runtime/jailer.go

```diff
@@ -0,0 +1,403 @@
+	ExposeDeviceToJail(string) error
+}
+
+// noopJailer is a jailer that returns only successful responses and performs
+// no operations during calls
+type noopJailer struct{}
+
+func (j noopJailer) BuildJailedMachine(cfg *Config, machineConfig *firecracker.Config, vmID string) ([]firecracker.Opt, error) {
+	return []firecracker.Opt{}, nil
+}
+
+func (j noopJailer) ExposeDeviceToJail(path string) error {
+	return nil
+}
+
+type runcJailConfig struct {
+	OCIBundlePath  string
+	UID            uint32
+	GID            uint32
+	RuncBinaryPath string
+}
+
+// runcJailer uses runc to set up a jailed environment for the Firecracker VM.
+type runcJailer struct {
+	logger *logrus.Entry
+
+	Config runcJailConfig
+}
+
+func newRuncJailer(logger *logrus.Entry, config runcJailConfig) *runcJailer {
+	return &runcJailer{
+		logger: logger,
+		Config: config,
+	}
+}
+
+// JailPath returns the base directory from where the jail binary will be ran
+// from
+func (j runcJailer) OCIBundlePath() string {
+	return j.Config.OCIBundlePath
+}
+
+// RootPath returns the root fs of the jailed system.
+func (j runcJailer) RootPath() string {
+	return filepath.Join(j.OCIBundlePath(), rootfsFolder)
+}
+
+func (j runcJailer) ContentsPath() string {
+	return filepath.Join(j.RootPath(), "var", "lib", "firecracker-containerd")
+}
+
+func (j *runcJailer) BuildJailedMachine(cfg *Config, machineConfig *firecracker.Config, vmID string) ([]firecracker.Opt, error) {
+	handler := j.BuildHandler(cfg, &machineConfig.SocketPath, vmID)
+	fifoHandler := j.BuildLinkFifoHandler()
+	client := firecracker.NewClient(machineConfig.SocketPath, j.logger, machineConfig.Debug)
+
+	opts := []firecracker.Opt{
+		firecracker.WithProcessRunner(j.jailerCommand(vmID)),
+		firecracker.WithClient(client),
+		func(m *firecracker.Machine) {
+			m.Handlers.FcInit = m.Handlers.FcInit.Prepend(handler)
+			m.Handlers.FcInit = m.Handlers.FcInit.AppendAfter(firecracker.CreateLogFilesHandlerName, fifoHandler)
+		},
+	}
+
+	return opts, nil
+}
+
+// BuildHandler will link the necessary files except for the fifos due to the
+// fifos needing to be created. Also, this will create the proper device nodes
+// needed by Firecracker
+func (j *runcJailer) BuildHandler(cfg *Config, socketPath *string, vmID string) firecracker.Handler {
+	ociBundlePath := j.OCIBundlePath()
+	rootPath := j.RootPath()
+	*socketPath = filepath.Join(rootPath, "api.socket")
+
+	return firecracker.Handler{
+		Name: jailerHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			const mode = os.FileMode(0700)
+			// Create the proper paths needed for the runc jailer
+			j.logger.WithField("rootPath", rootPath).Debug("Creating root drive path")
+			if err := os.MkdirAll(rootPath, mode); err != nil {
+				return errors.Wrapf(err, "failed to create root path: %v", rootPath)
+			}
+
+			binPath := filepath.Join(rootPath, "usr", "local", "bin")
+			j.logger.WithField("binPath", binPath).Debug("Creating /usr/local/bin")
```

Comment:
There's really no need to bother with paths like `/usr/local/bin` inside the jail. The various binaries can just live in the root directory.